homeassistant/components/diagnostics/__init__.py

```diff
@@ -84,24 +97,49 @@ async def get(  # pylint: disable=no-self-use
         self, request: web.Request, d_type: str, d_id: str
     ) -> web.Response:
         """Download diagnostics."""
-        if d_type != "config_entry":
+        if d_type not in D_TYPES:
             return web.Response(status=404)
 
         hass = request.app["hass"]
-        config_entry = hass.config_entries.async_get_entry(d_id)
 
-        if config_entry is None:
-            return web.Response(status=404)
+        if d_type == D_TYPE_CONFIG_ENTRY:
+            config_entry = hass.config_entries.async_get_entry(d_id)
 
-        info = hass.data[DOMAIN].get(config_entry.domain)
+            if config_entry is None:
+                return web.Response(status=404)
 
-        if info is None:
-            return web.Response(status=404)
+            info = hass.data[DOMAIN].get(config_entry.domain)
 
-        if info["config_entry"] is None:
-            return web.Response(status=404)
+            if info is None or info[D_TYPE_CONFIG_ENTRY] is None:
+                return web.Response(status=404)
+
+            data = await info[D_TYPE_CONFIG_ENTRY](hass, config_entry)
+            filename = f"{config_entry.domain}-{config_entry.entry_id}"
+        else:
+            dev_reg = async_get(hass)
+            device = dev_reg.async_get(d_id)
+
+            if device is None:
+                return web.Response(status=404)
+
+            func_list = {
+                info[D_TYPE_DEVICE]
+                for config_entry in async_config_entries_for_device(hass, device)
```

Comment:
Should we allow specifying the config entry + device combination instead? 

I can't think of a use case where a user wants to have debug information about both Unifi and Z-Wave JS.

Reply:
So the url for a device could be `/diagnostics/config_entry/config_entry_id/device_id`

Reply:
when you are on the device list coming from an integration card, the URL includes the config entry ID, but when you click into a device, the config entry is no longer in the URL. So I guess it depends on how we want to expose this in the UI?

Reply:
if the action to download the device diagnostic data is on the device page, would we have separate actions for each linked config entry?

Reply:
I am planning on re-doing the UI once we add device support to be a dialog. Users will be able to pick config entry / filter by device.

Reply:
> So the url for a device could be `/diagnostics/config_entry/config_entry_id/device_id`

should we make it `/diagnostics/config_entry/config_entry_id/device/device_id`? As in would there be any other diagnostic types that lived under a config entry?

Reply:
Sure that's fine. It doesn't really matter, it would need a frontend upgrade anyway